images.CI/measure-provisioners-duration.ps1

```diff
@@ -0,0 +1,121 @@
+            Command = $Command
+            StartTime = [DateTime]::Parse($DateTime)
+        }
+    }
+}
+
+function Invoke-TryFindProvisionerSubItem {
+    param([object]$Provisioner, [string] $Line)
+
+    if ($Provisioner.ProvisionerType -in "powershell", "shell", "windows-shell") {
+        if ($Line -match $ShellScriptSubItemRegex) {
+            Add-ProvisionerSubItem -Provisioner $Provisioner -Command $Matches[2] -DateTime $Matches[1]
+        }
+    } elseif ($Provisioner.ProvisionerType -eq "file") {
+        if ($Line -match $DownloadUploadSubItemRegex) {
+            Add-ProvisionerSubItem -Provisioner $Provisioner -Command $Matches[2] -DateTime $Matches[1]
+        }
+    }
+}
+
+function Assert-StartProvisioner {
+    param([object]$Provisioner, [string]$ProvisionerType)
+    if ($null -ne $Provisioner) {
+        throw "New provisioner '$ProvisionerType' has been started but previous '$($Provisioner.ProvisionerType)' was not finished yet"
+    }
+}
+
+function Assert-EndProvisioner {
+    param([object]$Provisioner, [string]$ProvisionerType)
+    if (($null -ne $Provisioner) -and ($Provisioner.ProvisionerType -ne $ProvisionerType)) {
+        throw "Expected end of '$($Provisioner.ProvisionerType)' provisioner but found end of '$ProvisionerType'"
+    }
+}
+
+$provisionersList = @()
+$currentProvisioner = $null
+
+if ((Get-Content $PackerLogPath -Raw) -notmatch $TelemetryLineRegex) {
+    throw "Packer log doesn't contain diagnostic information. Env PACKER_LOG must be set to 1"
+}
+ 
+Get-Content $PackerLogPath | ForEach-Object {
+    if ($_ -match $StartProvisionerRegex) {
+        Assert-StartProvisioner -Provisioner $currentProvisioner -ProvisionerType $Matches[2]
+
+        $currentProvisioner = New-ProvisionerItem -ProvisionerType $Matches[2] -StartTime $Matches[1]
+    } elseif (($_ -match $EndProvisionerRegex) -and $currentProvisioner) {
+        Assert-EndProvisioner -Provisioner $currentProvisioner -ProvisionerType $Matches[2]
+
+        Complete-ProvisionerItem -Provisioner $currentProvisioner -EndTime $Matches[1]
+        Add-ProvisionerSubItem -Provisioner $currentProvisioner -Command $null -DateTime $Matches[1]
+        $provisionersList += $currentProvisioner
+        $currentProvisioner = $null
+    } elseif ($currentProvisioner) {
+        Invoke-TryFindProvisionerSubItem -Provisioner $currentProvisioner -Line $_
+    }
+}
+$totalProvisionersTime = New-TimeSpan
+$provisionersList | ForEach-Object { $totalProvisionersTime = $totalProvisionersTime.Add($_.Duration) }
+
+# Print information about provisioners in order of execution
+Write-Host "Build timeline:"
+$provisionersList | ForEach-Object {
+    Write-Host "- $($_.Duration) | $($_.ProvisionerType)"
+    $_.SubItems | ForEach-Object {
+        Write-Host "      $($_.Duration) | $($_.Command)"
+    }
+    Write-Host ""
+}
+Write-Host "Total provisioners time: $totalProvisionersTime"
+
+if ($PrintTopNLongest -and ($PrintTopNLongest -gt 0)) {
```

Comment:
nitpick: remove `$PrintTopNLongest -and` and ()

Reply:
Thank you, fixed!